File: round_robin.py
#!/usr/bin/env python3
"""Round Robin Scheduling"""

def round_robin(processes, time_quantum):
    """processes: list of (pid, arrival_time, burst_time)"""
    procs = sorted(processes, key=lambda p: p[1])
    time = 0
    ready = []
    remaining = []
    completion_time = {}
    
    # track remaining burst time
    for pid, arrival, burst in procs:
        remaining.append([pid, arrival, burst, burst])  # [pid, arrival, original_burst, remaining_burst]
    
    queue = []
    idx = 0
    
    while queue or idx < len(remaining):
        # add newly arrived processes
        while idx < len(remaining) and remaining[idx][1] <= time:
            queue.append(remaining[idx])
            idx += 1
        
        if queue:
            current = queue.pop(0)
            pid, arrival, original_burst, rem_burst = current
            
            exec_time = min(time_quantum, rem_burst)
            time += exec_time
            rem_burst -= exec_time
            
            # add newly arrived during execution
            while idx < len(remaining) and remaining[idx][1] <= time:
                queue.append(remaining[idx])
                idx += 1
            
            if rem_burst > 0:
                current[3] = rem_burst
                queue.append(current)
            else:
                completion_time[pid] = time
        else:
            # idle
            if idx < len(remaining):
                time = remaining[idx][1]
    
    # build result
    result = []
    for pid, arrival, burst, _ in remaining:
        ct = completion_time[pid]
        tat = ct - arrival
        wt = tat - burst
        result.append((pid, arrival, burst, ct, tat, wt))
    
    result.sort(key=lambda r: r[0])  # sort by pid
    return result


def print_table(result):
    print("{:>4} {:>4} {:>4} {:>4} {:>5} {:>5}".format("PID", "AT", "BT", "CT", "TAT", "WT"))
    for row in result:
        print("{:>4} {:>4} {:>4} {:>4} {:>5} {:>5}".format(*row))
    n = len(result)
    avg_tat = sum(r[4] for r in result) / n
    avg_wt = sum(r[5] for r in result) / n
    print(f"\nAverage Turnaround Time = {avg_tat:.2f}")
    print(f"Average Waiting Time    = {avg_wt:.2f}")


if __name__ == '__main__':
    processes = [
        (1, 0, 5),
        (2, 1, 3),
        (3, 2, 1),
        (4, 3, 2),
    ]
    quantum = 2
    res = round_robin(processes, quantum)
    print(f"Round Robin Scheduling (Time Quantum = {quantum}):\n")
    print_table(res)

File: round_robin.js
#!/usr/bin/env node
/**
 * Round Robin Scheduling
 */

function roundRobin(processes, timeQuantum) {
    const procs = [...processes].sort((a, b) => a.arrival - b.arrival);
    let time = 0;
    const queue = [];
    const completionTime = {};
    
    // track remaining burst time
    const remaining = procs.map(p => ({
        ...p,
        remainingBurst: p.burst
    }));
    
    let idx = 0;
    
    while (queue.length > 0 || idx < remaining.length) {
        // add newly arrived processes
        while (idx < remaining.length && remaining[idx].arrival <= time) {
            queue.push(remaining[idx]);
            idx++;
        }
        
        if (queue.length > 0) {
            const current = queue.shift();
            
            const execTime = Math.min(timeQuantum, current.remainingBurst);
            time += execTime;
            current.remainingBurst -= execTime;
            
            // add newly arrived during execution
            while (idx < remaining.length && remaining[idx].arrival <= time) {
                queue.push(remaining[idx]);
                idx++;
            }
            
            if (current.remainingBurst > 0) {
                queue.push(current);
            } else {
                completionTime[current.pid] = time;
            }
        } else {
            // idle
            if (idx < remaining.length) {
                time = remaining[idx].arrival;
            }
        }
    }
    
    // build result
    const result = procs.map(p => {
        const ct = completionTime[p.pid];
        const tat = ct - p.arrival;
        const wt = tat - p.burst;
        return {
            pid: p.pid,
            arrival: p.arrival,
            burst: p.burst,
            completion: ct,
            turnaround: tat,
            waiting: wt
        };
    });
    
    result.sort((a, b) => a.pid - b.pid);
    return result;
}

function printTable(result) {
    console.log("PID".padStart(4), "AT".padStart(4), "BT".padStart(4), 
                "CT".padStart(4), "TAT".padStart(5), "WT".padStart(5));
    
    for (const row of result) {
        console.log(
            String(row.pid).padStart(4),
            String(row.arrival).padStart(4),
            String(row.burst).padStart(4),
            String(row.completion).padStart(4),
            String(row.turnaround).padStart(5),
            String(row.waiting).padStart(5)
        );
    }
    
    const n = result.length;
    const avgTAT = result.reduce((sum, r) => sum + r.turnaround, 0) / n;
    const avgWT = result.reduce((sum, r) => sum + r.waiting, 0) / n;
    
    console.log(`\nAverage Turnaround Time = ${avgTAT.toFixed(2)}`);
    console.log(`Average Waiting Time    = ${avgWT.toFixed(2)}`);
}

// Main execution
if (require.main === module) {
    const processes = [
        { pid: 1, arrival: 0, burst: 5 },
        { pid: 2, arrival: 1, burst: 3 },
        { pid: 3, arrival: 2, burst: 1 },
        { pid: 4, arrival: 3, burst: 2 }
    ];
    
    const quantum = 2;
    const result = roundRobin(processes, quantum);
    console.log(`Round Robin Scheduling (Time Quantum = ${quantum}):\n`);
    printTable(result);
}

module.exports = { roundRobin, printTable };
